src/solver.rs

```diff
@@ -124,79 +128,215 @@ pub fn resolve<P: Package, V: Version>(
-                .iter()
-                .any(|incompat| state.is_terminal(incompat))
-            {
-                // For a dependency incompatibility to be terminal,
-                // it can only mean that root depend on not root?
-                return Err(PubGrubError::Failure(
-                    "Root package depends on itself at a different version?".into(),
-                ));
+                        dependent: dependent.clone(),
+                    });
+                }
+
+                x
             }
-            state.partial_solution.add_version(
-                p.clone(),
-                v,
-                dep_incompats,
-                &state.incompatibility_store,
-            );
-        } else {
+        };
+
+        // Add that package and version if the dependencies are not problematic.
+        let dep_incompats =
+            state.add_incompatibility_from_dependencies(p.clone(), v.clone(), &dependencies);
+
+        // TODO: I don't think this check can actually happen.
+        // We might want to put it under #[cfg(debug_assertions)].
+        if state.incompatibility_store[dep_incompats.clone()]
+            .iter()
+            .any(|incompat| state.is_terminal(incompat))
+        {
+            // For a dependency incompatibility to be terminal,
+            // it can only mean that root depend on not root?
+            return Err(PubGrubError::Failure(
+                "Root package depends on itself at a different version?".into(),
+            ));
+        }
+
+        state.partial_solution.add_version(
+            p.clone(),
+            v,
+            dep_incompats,
+            &state.incompatibility_store,
+        );
+    }
+}
+
+/// Main function of the library.
+/// Finds a set of packages satisfying dependency bounds for a given package + version pair.
+#[cfg(feature = "async")]
+pub async fn resolve_async<P: Package, V: Version>(
+    dependency_provider: &impl AsyncDependencyProvider<P, V>,
+    package: P,
+    version: impl Into<V>,
+) -> Result<SelectedDependencies<P, V>, PubGrubError<P, V>> {
+    let mut state = State::init(package.clone(), version.into());
+    let mut added_dependencies: Map<P, Set<V>> = Map::default();
+    let mut next = package;
+    loop {
+        dependency_provider
+            .should_cancel()
+            .map_err(|err| PubGrubError::ErrorInShouldCancel(err))?;
+
+        state.unit_propagation(next)?;
+
+        let potential_packages = match state.partial_solution.potential_packages() {
+            Some(packages) => packages,
+            None => {
+                return state.partial_solution.extract_solution().ok_or_else(|| {
+                    PubGrubError::Failure(
+                        "How did we end up with no package to choose but no solution?".into(),
+                    )
+                });
+            }
+        };
+
+        let decision = dependency_provider
+            .choose_package_version(potential_packages)
```

Comment:
It would be really nice if the asynchronous calls to get package information can be run in parallel allowing this await to return when the first of them returns. Of course this would make reproducibility much more complicated, and would need to be documented carefully. But it should make many real-world cases much more efficient.